Feature/find by postcode - Postcode.io integration and error messages as well as journey

beseven opened the pull request:
<!--
Thanks for wanting to contribute to this repository.

In order for the changes to be integrated into the repo with as little friction
as possible please follow the guidance here. This includes completing all
sections as fully as possible.
Prior to creating a Pull Request, please ensure there is an open issue for the
changes you wish to make. This will provide visibility to others early in the
process. Potentially other people will wish to help out. It also allows us to
validate the change is inline with our vision for the product.

Provide a general summary of your changes in the Title
-->

## Description
Location page for sexual health experts (including query building)

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://jira.service.nhs.uk/browse/PCFC-171

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Part of the Find a chlamydia test journey
Fill out the journey from start to results

## Checklist
<!-- Ensure each of the points below have been considered and completed where applicable -->
- [ x ] New and/or updated tests
- [ x ] Add things to CHANGELOG

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -1,7 +1,51 @@
 }
 
+function location(req, res) {
+  if (res.locals.locationHeading === null) {
```
Using the fact `locationHeading` hasn't been set is a side effect of the location not being correct, this might change. It would be clearer to set a specific flag and use that to check if an error has occurred.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -1,7 +1,51 @@
+}
+
+function emptyPostcodePage(req, res) {
+  log.debug('Empty Search');
```
Logs are best used when there is additional information to capture. It would be better to do something like increment a Prometheus counter.

beseven replied:
Will keep that in mind when we add metrics to the app.

st3v3nhunt commented on test/integration/resultsPageErrors.js:
```diff
@@ -0,0 +1,61 @@
+
+function expectErrorMessage(res, message) {
+  const $ = cheerio.load(res.text);
+  log.debug(cheerio.load(res.text));
```
Should this be here?

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -1,7 +1,51 @@
 module.exports = {
   findHelp,
+  location,
+  emptyPostcodePage,
```
There is no need to have the name of the view being rendered suffixed with `page`. The same goes for `invalidPostcodePage` and `outsideOfEnglandPage`.

st3v3nhunt commented on app/middleware/locationValidator.js:
```diff
@@ -0,0 +1,13 @@
+const renderer = require('../middleware/renderer');
+
+function validateLocation(req, res, next) {
+  const postcodeSearch = res.locals.location;
```
There is no need for a variable here, if can be combined within the `if`.

st3v3nhunt commented on app/middleware/postcodeLookup.js:
```diff
@@ -0,0 +1,52 @@
+}
+
+async function lookupPostcode(req, res, next) {
+  const postcodeSearch = res.locals.location;
```
Does the variable name need to be changed here? I'd prefer to see `const location = res.locals.location;` for ease of understanding.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -1,7 +1,51 @@
+  next(error);
+}
+
+function invalidPostcodePage(postcode, req, res) {
```
Can the parameter name be `location` rather than `postcode`? It makes more sense when viewed with the log statement which is `Location failed...`.
Also, for consistency with the other functions it would be good have the parameter list begin with `req, res, ...`.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -1,7 +1,51 @@
+  location(req, res);
+}
+
+function outsideOfEnglandPage(loc, req, res) {
```
For consistency with the other functions it would be good have the parameter list begin with `req, res, ...`.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -1,7 +1,51 @@
+  location(req, res);
+}
+
+function postcodeError(error, postcode, res, next) {
```
For consistency with the other functions it would be good have the parameter list begin with `req, res, ...`.
Also, this function doesn't render a view within the function as it delegates that to the global error handling by passing an error into `next`. I suggest this function is removed and the logging and next call are moved back to current caller of this function. I don't setting `res.locals.errorMessage` does anything in this instance either as the error view has no use of that variable.

st3v3nhunt commented on test/integration/locationPage.js:
```diff
@@ -0,0 +1,96 @@
+const chai = require('chai');
```
Good, thorough tests 👍

beseven replied:
Picked up from @neilbmclaughlin's work :)

st3v3nhunt commented on test/integration/resultsPageErrors.js:
```diff
@@ -0,0 +1,61 @@
+const cheerio = require('cheerio');
```
Can all external dependencies be listed before internal modules please? Ideally alphabetically.

st3v3nhunt commented on test/unit/middleware/renderer.js:
```diff
@@ -0,0 +1,102 @@
+    });
+  });
+
+  describe('results', () => {
```
Worth taking this out until there are some results?

beseven commented on app/middleware/setLocals.js:
```diff
@@ -0,0 +1,28 @@
+  res.locals.type = req.query.type;
+  res.locals.origin = req.query.origin;
+  res.locals.location = req.query.location;
+  if ((req.query.type === constants.SERVICE_TYPES.professional)
```
If I do this, then I will have to link the correctParams flag to the setting of the locationHeading. You can take a look at re-review

beseven commented on app/middleware/setLocals.js:
```diff
@@ -0,0 +1,28 @@
+const constants = require('../lib/constants');
```
In order to not pass the res object around, kept the constants.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -15,7 +18,36 @@ function age(req, res) {
+}
+
+function invalidPostcode(req, res, loc) {
+  log.debug({ loc }, 'Location failed validation');
```
Can the debug payload be changed to `{ location: loc }` so that there is consistency in field for which the `location` search term is recorded against please.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -15,7 +18,36 @@ function age(req, res) {
+}
+
+function outsideOfEngland(req, res, loc) {
+  log.debug({ loc }, 'Outside of England');
```
`loc` is the response object from Postcodes.io. Can this be made clear or if that isn't the intention (as it has already been logged in the [postcodeLookup](https://github.com/nhsuk/sexual-health-service-finder/pull/11/files#diff-ea9d7600f9dd35965091987f6fa94e04R38) can jus the location of the request be logged (that isn't included in Postcodes.io response).

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -15,7 +18,36 @@ function age(req, res) {
+
+function location(req, res) {
+  if (!res.locals.correctLocationParams) {
+    return res.render('error');
```
The `error` view contains the message about there having been a technical fault. As that isn't what has happened can this render a (possibly new) view informing the user they need to start over. Or, as I think has been agreed previously, can the start page be returned. However, if that does happen, there might need to be some way of informing the user what has happened.

beseven replied:
Suggest to send to the `start` page and have an error message displayed.

st3v3nhunt commented on app/middleware/renderer.js:
```diff
@@ -15,7 +18,36 @@ function age(req, res) {
+}
+
+function emptyPostcode(req, res) {
+  log.debug('Empty search');
```
Can this be replaced with a prometheus counter?

st3v3nhunt commented on app/views/error.nunjucks:
```diff
@@ -11,4 +11,4 @@
     <p>Please try again later.</p>
   <div>
-{% endblock %}
+{% endblock %}
```
New line please.

st3v3nhunt commented on config/routes.js:
```diff
@@ -2,6 +2,10 @@ const router = require('express').Router();
 const renderer = require('../app/middleware/renderer');
 const selectionValidatorSymptoms = require('../app/middleware/selectionValidatorSymptoms');
 const selectionValidatorAge = require('../app/middleware/selectionValidatorAge');
+const setLocals = require('../app/middleware/setLocals');
```
Can the require's be ordered alphabetically please.

st3v3nhunt commented on config/routes.js:
```diff
@@ -10,24 +14,43 @@ router.get(
 
 router.get(
   '/symptoms',
+  setLocals.fromRequest,
```
If `setLocals` is merged with `locals` there is no need to have this here.

st3v3nhunt commented on config/routes.js:
```diff
@@ -10,24 +14,43 @@ router.get(
 
 router.get(
   '/recommend',
+  setLocals.fromRequest,
```
If `setLocals` is merged with `locals` there is no need to have this here.

st3v3nhunt commented on config/routes.js:
```diff
@@ -10,24 +14,43 @@ router.get(
 
 router.get(
   '/age',
+  setLocals.fromRequest,
```
If `setLocals` is merged with `locals` there is no need to have this here.

st3v3nhunt commented on config/routes.js:
```diff
@@ -10,24 +14,43 @@ router.get(
 
 router.get(
   '/choose',
+  setLocals.fromRequest,
```
If `setLocals` is merged with `locals` there is no need to have this here.

st3v3nhunt commented on config/routes.js:
```diff
@@ -10,24 +14,43 @@ router.get(
 
+router.get(
+  '/location',
+  setLocals.fromRequest,
```
If `setLocals` is merged with `locals` there is no need to have this here.

st3v3nhunt commented on config/routes.js:
```diff
@@ -10,24 +14,43 @@ router.get(
+
+router.get(
+  '/results',
+  setLocals.fromRequest,
```
If `setLocals` is merged with `locals` there is no need to have this here.

st3v3nhunt commented on app/middleware/setLocals.js:
```diff
@@ -0,0 +1,64 @@
+
+module.exports = {
+  fromRequest,
+  getLocationHeading,
```
This function isn't used outside of the module, please remove it from the exports.

st3v3nhunt commented on app/middleware/setLocals.js:
```diff
@@ -0,0 +1,64 @@
+  fromRequest,
+  getLocationHeading,
+  mapServiceType,
+  mapServiceChoice,
```
See comment above.

st3v3nhunt commented on app/middleware/setLocals.js:
```diff
@@ -0,0 +1,64 @@
+module.exports = {
+  fromRequest,
+  getLocationHeading,
+  mapServiceType,
```
See comment above.

st3v3nhunt commented on app/middleware/setLocals.js:
```diff
@@ -0,0 +1,64 @@
+
+function fromRequest(req, res, next) {
+  res.locals.locationHeading = null;
+  res.locals.locationLabel = 'Enter a postcode in England.';
```
`locationLabel` is only used in `location.nunjucks` - does it need to be set as a local?

beseven replied:
No, not really

beseven commented on app/middleware/postcodeLookup.js:
```diff
@@ -0,0 +1,57 @@
+        validationCounter.inc(1);
+      }
+    } catch (error) {
+      log.debug({ location }, 'Error in location');
```
`Error in postcode lookup`

beseven commented:
Hey @st3v3nhunt, I've addressed all your concerns bar the `nock` one. I'm happy to raise an issue and refactor at a later date, based on priority. Is that ok?

st3v3nhunt commented on app/middleware/locals.js:
```diff
@@ -1,3 +1,49 @@
+  return undefined;
+}
+
 // eslint-disable-next-line no-unused-vars
```
I think this line can be removed now.

st3v3nhunt commented on app/middleware/locals.js:
```diff
@@ -1,3 +1,49 @@
+const constants = require('../lib/constants');
+
+function getLocationHeading(query) {
```
Can this function be moved into a separate module?
It would facilitate some unit tests (I know there are integration tests already so it might be ok to leave as is) and means this module is just about setting locals.

st3v3nhunt commented on app/middleware/locals.js:
```diff
@@ -1,3 +1,49 @@
+  return undefined;
+}
+
+function mapServiceType(query) {
```
Can this function be moved into a separate module?
It would facilitate some unit tests (I know there are integration tests already so it might be ok to leave as is) and means this module is just about setting locals.

st3v3nhunt commented on app/middleware/locals.js:
```diff
@@ -1,3 +1,49 @@
+  return undefined;
+}
+
+function mapServiceChoice(query) {
```
Can this function be moved into a separate module?
It would facilitate some unit tests (I know there are integration tests already so it might be ok to leave as is) and means this module is just about setting locals.

st3v3nhunt commented on app/lib/utils/queryMapper.js:
```diff
@@ -0,0 +1,52 @@
+  mapServiceChoice,
+  mapServiceType,
+};
+
```
Blank line can be removed please.
